hail/python/hailtop/utils/utils.py

```diff
@@ -26,34 +26,93 @@ def async_to_blocking(coro):
         thread_pool, lambda: fun(*args, **kwargs))
 
 
-class AsyncWorkerPool:
-    def __init__(self, parallelism):
-        self._sem = asyncio.Semaphore(parallelism)
-        self._count = 0
+async def gather(*coros, parallelism=10, return_exceptions=False):
+    gatherer = AsyncThrottledGather(*coros,
+                                    parallelism=parallelism,
+                                    return_exceptions=return_exceptions)
+    return await gatherer.wait()
+
+
+class AsyncThrottledGather:
+    def __init__(self, *coros, parallelism=10, return_exceptions=False):
+        self.count = len(coros)
+        self.n_finished = 0
+
+        self._queue = asyncio.Queue(maxsize=1000)
         self._done = asyncio.Event()
+        self._return_exceptions = return_exceptions
+        self._futures = []
+
+        self._results = [None] * len(coros)
+        self._error = None
+
+        if self.count == 0:
+            self._done.set()
+            return
+
+        for _ in range(parallelism):
+            self._futures.append(asyncio.ensure_future(self._worker()))
+
+        self._futures.append(asyncio.ensure_future(self._fill_queue(*coros)))
+
+    def _cancel_futures(self):
+        for fut in self._futures:
+            fut.cancel()
+
+    async def _worker(self):
+        while True:
+            i, coro = await self._queue.get()
 
-    async def _call(self, f, args, kwargs):
-        async with self._sem:
             try:
-                await f(*args, **kwargs)
+                res = await coro
```

Comment:
I think my way is clearer. The behavior of this function should insert the exception as the result for that coroutine rather than raising an exception if raise_exceptions is False.